### ISSUE TITLE
Add FFTW3 package

### ISSUE THREAD

**Issue by lukas-w** [ruslo/hunter#1160]
Website: http://www.fftw.org/
GitHub: https://github.com/FFTW/fftw3

FFTW3 has a CMake build system since its last release. It offers multiple options at build time:

* `ENABLE_FLOAT`, `ENABLE_LONG_DOUBLE`, `ENABLE_QUAD_PRECISION` (mutually exclusive). This sets the floating point precision to use. The resulting library will be suffixed with `f`, `l`, `q` respectively. I'm not sure how to provide this option through Hunter for the user. Should we just offer multiple packages (`fftw3`, `fftw3f`, `fftw3l` and `fftw3q`) [the way Ubuntu does it](https://packages.ubuntu.com/source/trusty/fftw3)?
* `ENABLE_SSE`, `ENABLE_SSE2`, `ENABLE_AVX`, `ENABLE_AVX2`. These enable support for the respective instruction sets. Should we call [`hunter_check_toolchain_definition`](https://docs.hunter.sh/en/latest/reference/user-modules/hunter_check_toolchain_definition.html) to detect support and enable the options accordingly?
* `ENABLE_OPENMP`, `ENABLE_THREADS` (pthreads) enable multi-threading support. They're mutually exclusive and will result in a `_omp` or `_threads` suffix. How should this option be exposed to the user?

**Comment by ruslo** [comment]
> I'm not sure how to provide this option through Hunter for the user

See https://github.com/ruslo/hunter/wiki/example.custom.config.id#custom-cmake-options

> Should we just offer multiple packages

Don't know all functionality details but I don't think so.

> Should we call hunter_check_toolchain_definition to detect support and enable the options accordingly?

`hunter_check_toolchain_definition` is a helper to analyze toolchain. Toolchain (compiler, flags) provided by user. If that is what is expected then yes, it can be used.

> How should this option be exposed to the user?

By `CMAKE_ARGS`, see above.

**Comment by lukas-w** [comment]
Thanks for pointing me in the right direction. Will the `CMAKE_ARGS` solution work with the binary cache servers? I guess only the default option will be available cached?

**Comment by ruslo** [comment]
> Will the CMAKE_ARGS solution work with the binary cache servers? I guess only the default option will be available cached?

You can add default `CMAKE_ARGS` in `hunter.cmake`:
* https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#cmake-options

These options will be used while testing and uploading. You can add more examples with different `CMAKE_ARGS` and add them to testing matrix, e.g.:
* https://github.com/ruslo/hunter/blob/78c5d57868ebe622f08ab1fb4334b2d3e9f877a5/examples/OpenCV-extra/config.cmake#L1-L9
* https://github.com/ruslo/hunter/blob/78c5d57868ebe622f08ab1fb4334b2d3e9f877a5/examples/OpenCV-extra/CMakeLists.txt#L5

> I can't find any documentation on how to specify components, but I guess I'll be able to learn about it by reading the existing packages' scripts.

What exactly do you mean by term "components"? As far as I can see you're talking about different targets so it purely about CMake code of the package. The `hunter_add_package(... COMPONENTS ...)` feature used only for non-CMake build schemes. At this moment it's used only in Qt and Boost.

**Comment by lukas-w** [comment]
> What exactly do you mean by term "components"?

Yes, I'm talking about the `hunter_add_package(... COMPONENTS ...) ` feature. As far as I understand the source code, the `COMPONENTS` option simply includes subdirectories' `hunter.cmake` files, so `hunter_add_package(foo COMPONENTS bar)` will include `projects/foo/hunter.cmake` and `projects/foo/bar/hunter.cmake`. 

Using `CMAKE_ARGS` is not enough because users may want e.g. `fftw3f` and `fftw3l` side-by-side, but FFTW's `CMakeLists.txt` only allows building one of those at a a time. So I think components are the way to go here. I guess we'll need a layout like this:

```
./projects/fftw3
├── fftw3
|   ├── hunter.cmake
├── fftw3f
|   └── hunter.cmake
├── fftw3l
|   └── hunter.cmake
├── fftw3q
|   └── hunter.cmake
└── hunter.cmake
```

Each subdirectory's `hunter.cmake` would look something like this, but differ in the supplied CMake args and the `PACKAGE_NAME`:
```cmake
include(hunter_add_version)
include(hunter_pick_scheme)
include(hunter_download)
include(hunter_cacheable)
include(hunter_cmake_args)

hunter_add_version(
    PACKAGE_NAME fftw3
    VERSION "3.3.7"
    URL "http://www.fftw.org/fftw-3.3.7.tar.gz"
    SHA1 2ae980a8d44c161ce4a09c6e2d1c79243ecbabb2
)

hunter_cmake_args(fftw3 CMAKE_ARGS
    BUILD_TEST=OFF
#  ENABLE_FLOAT=ON
#  ENABLE_LONG_DOUBLE=ON
#  ....
)

hunter_pick_scheme(DEFAULT url_sha1_cmake)
hunter_cacheable(fftw3)
hunter_download(PACKAGE_NAME fftw3)
```

Is there a way to avoid the source archive to be downloaded multiple times?

**Comment by ruslo** [comment]
> Yes, I'm talking about the hunter_add_package(... COMPONENTS ...) feature

As I said before this feature is for non-CMake builds only, it's not our case as far as I understand.

> users may want e.g. fftw3f and fftw3l side-by-side

If there is only one archive with sources then it's a one package. If there are sources for `fftw3f` and `fftw3l` and **there is no intersection** in installed files then it's simply two different packages which can be installed by `hunter_add_package(fftw3f)` and `hunter_add_package(fftw3l)`. And again, there is no need for `COMPONENTS` as you see.

**Comment by lukas-w** [comment]
> As I said before this feature is for non-CMake builds only, it's not our case as far as I understand.

It's only used in non-CMake builds right now, but couldn't it be useful for CMake builds as well? 

> If there is only one archive with sources then it's a one package. 

There's only one archive with sources, but it only supports building one version of the library at a time. So if we wanted `fftw3f` and `ffwt3l` we'd still need a custom scheme to invoke CMake configure/build/install steps multiple times with different `CMAKE_ARGS`. 

> If there are sources for fftw3f and fftw3l and there is no intersection

There is an intersection of installed files, `include/fftw3.h`.

So none of the ways we've discussed so far appear to be a good solution. Maybe the cleanest way is to patch upstream's `CMakeLists.txt` so it can build all versions at once and have multiple targets?

**Comment by ruslo** [comment]
> it only supports building one version of the library at a time

It means this feature should be introduced by package itself.

> There is an intersection of installed files, include/fftw3.h

Then it's the same package

> Maybe the cleanest way is to patch upstream's CMakeLists.txt so it can build all versions at once and have multiple targets?

Yes.

**Comment by lukas-w** [comment]
Great, I'll work on getting a PR ready to submit upstream. 👍 Thanks for your help.

**Comment by ruslo** [comment]
I've created a fork if you need a temporary place for experiments/hosting release archives:
* https://github.com/hunter-packages/fftw3

We can remove it when feature will be available upstream.

@lukas-w see https://docs.hunter.sh/en/latest/creating-new/patch.html
